File: ruby/calculator_cli_app/src/queries/calculation_query.rb
# frozen_string_literal: true
# rbs_inline: enabled

require 'net/http'
require 'json'

module Queries
  class CalculationQuery
    # @rbs seed: String
    # @rbs return: void
    def initialize(seed)
      @seed = seed
    end

    # @rbs num: Integer
    # @rbs return: Integer
    def f(num)
      if num.zero?
        1
      elsif num == 2
        2
      elsif num.even?
        f(num - 1) + f(num - 2) + f(num - 3) + f(num - 4)
      else
        ask_server(num)
      end
    end

    private

    attr_reader :seed

    # @rbs return: URI::Generic
    def uri
      @uri ||= URI.parse(ENV.fetch('CALCULATION_API'))
    end

    # @rbs num: Integer
    # @rbs return: Integer
    def ask_server(num)
      uri.query = URI.encode_www_form(seed:, n: num)
      req       = Net::HTTP::Get.new(uri)
      res       = Net::HTTP.start(uri.host.to_s, uri.port) { |http| http.request(req) }
      JSON.parse(res.body)['result']
    end
  end
end

File: javascript/calculator_cli_app/src/queries/calculation_query.js
const http = require("http")

const CalculationQuery = class {
  constructor(seed) {
    this.seed = seed
    this.uri  = null
  }

  f(n) {
    if (n == 0) {
      return 1;
    } else if (n == 2) {
      return 2;
    } else if (n % 2 == 0) {
      return this.f(n - 1) + this.f(n - 2) + this.f(n - 3) + this.f(n - 4)
    } else {
      return this.__ask_server__(n);
    }
  }

  // private

  __uri__() {
    if (this.uri === null) {
      this.uri = process.env.CALCULATION_API;
      return this.uri;
    } else {
      return this.uri;
    }
  }

  __ask_server__(n) {
    let result;
    http.get(`${this.__uri__()}?seed=${this.seed}&n=${n}`, (res) => {
      let body = "";
      res.setEncoding("utf8");

      res.on("data", (chunk) => {
        body += chunk;
      });

      res.on("end", (res) => {
        result = JSON.parse(body).result;
      });

    }).on("error", (error) => {
      result = error.message;
    });
    return result;
  }
}

export default CalculationQuery
